cmd/root.go

```diff
@@ -52,34 +56,91 @@ func newRunCommand(runnerConfig *actions.RunnerConfig) func(*cobra.Command, []st
 			runnerConfig.EventName = args[0]
 		}
 
-		// create the runner
-		runner, err := actions.NewRunner(runnerConfig)
+		err := parseAndRun(cmd, runnerConfig)
 		if err != nil {
 			return err
 		}
-		defer runner.Close()
-
-		// check if we should just print the graph
-		list, err := cmd.Flags().GetBool("list")
+		watch, err := cmd.Flags().GetBool("watch")
 		if err != nil {
 			return err
 		}
-		if list {
-			return drawGraph(runner)
+		if watch {
+			return watchAndRun(runnerConfig.Ctx, func() {
+				err = parseAndRun(cmd, runnerConfig)
+			})
 		}
 
-		// check if we are running just a single action
-		actionName, err := cmd.Flags().GetString("action")
-		if err != nil {
-			return err
-		}
-		if actionName != "" {
-			return runner.RunActions(actionName)
-		}
+		return err
+	}
+}
+
+func parseAndRun(cmd *cobra.Command, runnerConfig *actions.RunnerConfig) error {
+	// create the runner
+	runner, err := actions.NewRunner(runnerConfig)
+	if err != nil {
+		return err
+	}
+	defer runner.Close()
+
+	// check if we should just print the graph
+	list, err := cmd.Flags().GetBool("list")
+	if err != nil {
+		return err
+	}
+	if list {
+		return drawGraph(runner)
+	}
 
-		// run the event in the RunnerRonfig
-		return runner.RunEvent()
+	// check if we are running just a single action
+	actionName, err := cmd.Flags().GetString("action")
+	if err != nil {
+		return err
 	}
+	if actionName != "" {
+		return runner.RunActions(actionName)
+	}
+
+	// run the event in the RunnerRonfig
+	return runner.RunEvent()
+}
+
+func watchAndRun(ctx context.Context, fn func()) error {
+	recurse := true
+	checkIntervalInSeconds := 2
+	dir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	var ignore *gitignore.GitIgnore
+	if _, err := os.Stat(filepath.Join(dir, ".gitignore")); !os.IsNotExist(err) {
+		ignore, _ = gitignore.CompileIgnoreFile(filepath.Join(dir, ".gitignore"))
+	} else {
+		ignore = &gitignore.GitIgnore{}
+	}
+
+	folderWatcher := fswatch.NewFolderWatcher(
+		dir,
+		recurse,
+		ignore.MatchesPath,
+		checkIntervalInSeconds,
+	)
+
+	folderWatcher.Start()
+	log.Debugf("Watching %s for changes", dir)
+
+	go func() {
+		for folderWatcher.IsRunning() {
+			for changes := range folderWatcher.ChangeDetails() {
+				log.Debugf("%s", changes.String())
+				fn()
```

Comment:
the `fn` parameter for this `watchAndRun` function should be `fn func() error` and then use the return of that function to be the return for `watchAndRun` to avoid swallowing the error.